networklibrary/src/main/java/com/eis/communication/network/NetworkManager.java

```diff
@@ -0,0 +1,47 @@
+package com.eis.communication.network;
+
+import com.eis.communication.Peer;
+
+/**
+ * Manager of any type of network, contains methods that should be in every network
+ *
+ * @param <RK> Type of resource key handled by the network
+ * @param <RV> Type of resource value handled by the network
+ * @param <P>  Type of peer/address handled by the network
+ * @param <FR> The enumeration used to define fail reasons for the network.
+ * @author Luca Crema
+ * @author Marco Mariotto
+ * @author Alberto Ursino
+ * @author Alessandra Tonin
+ */
+public interface NetworkManager<RK, RV, P extends Peer, FR extends FailReason> {
+
+    /**
+     * Saves a resource value in the network for the specified key. If the save was successful
+     * {@link SetResourceListener#onResourceSet(Object, Object)} will be called.
+     *
+     * @param key                 The key identifier for the resource.
+     * @param value               The identified value of the resource.
+     * @param setResourceListener Listener called on resource successfully saved or on fail.
+     */
+    void setResource(RK key, RV value, SetResourceListener<RK, RV, FR> setResourceListener);
+
+    /**
+     * Retrieves a resource value from the network for the specified key. The value is returned inside
+     * {@link GetResourceListener#onGetResource(Object, Object)}.
+     *
+     * @param key                 The key identifier for the resource.
+     * @param getResourceListener Listener called on resource successfully retrieved or on fail.
+     */
+    void getResource(RK key, GetResourceListener<RK, RV, FR> getResourceListener);
```

Comment:
I have a few questions about the theoretical implementation of these interfaces; considering the different nature of the requests (invite, getResource, setResource) the listeners won't operate at the same way for the three requests, for example, the onResourceSet method will verify if the resource has been added (with a replicated dictionary it could mean verifying that each node has received the resource, but I guess that it's inevitable), the onGetResource method instead returns the searched resource; but you pass the listener as a parameter, so I imagine that when the method (for example) getResouce is called:
1. Is sent a request (in some format to be decided)
2. The listener is called, and it awaits for the response, and operate consequently

Reply:
If the request is to add a resource, the listener will wait for an acknowledge message that will notify him about the successful add, if the request is not sent/there is no answer from the receiving nodes (after a specified time?) the listener just calls the onResourceSetFail

Reply:
But these listeners' duty is just to verify the suxcessfully retrieved resource/add/invite, you didn't specify how we should handle the incoming messagges, am I right? To receive and process a setResource request from another node I will need to create another listener that will recognize the incoming messages

Reply:
I lost you at the last comment, what do you mean with
> To receive and process a setResource request from another node I will need to create another listener that will recognize the incoming messages
___
 Yes, I don't specify how the network handles messages/requests.
___
For the `getResource` method on a replicated dictionary, we could have used the return value of the method because it can be done synchronously and doesn't actually have any need for a listener, but I wanted to use the same interface for kademlia too, where a listener is needed because getResource is a request

Reply:
> For the getResource method on a replicated dictionary, we could have used the return value of the method because it can be done synchronously and doesn't actually have any need for a listener, but I wanted to use the same interface for kademlia too, where a listener is needed because getResource is a request
Yeah, that was not a proble, actually I liked the idea considering that we will have to adapt the KademliaNetwork to tha same interfaces.
And about my last comment I was starting to think how to handle messages and requests, but as you said you don't specify how to do it, so I guess that it's a problem of mine, as long as it doesn't directly depend on these interfaces. So, problem solved